ledger/acctupdates.go

```diff
@@ -246,51 +248,16 @@ func (au *accountUpdates) close() {
 	au.waitAccountsWriting()
 }
 
-func (au *accountUpdates) lookup(rnd basics.Round, addr basics.Address, withRewards bool) (data basics.AccountData, err error) {
+// Lookup returns the accound data for a given address at a given round. The withRewards indicates whether the
+// rewards should be added to the AccountData before returning. Note that the function doesn't update the account with the rewards,
+// even while it could return the AccoutData which represent the "rewarded" account data.
+func (au *accountUpdates) Lookup(rnd basics.Round, addr basics.Address, withRewards bool) (data basics.AccountData, err error) {
 	au.accountsMu.RLock()
 	defer au.accountsMu.RUnlock()
-	offset, err := au.roundOffset(rnd)
-	if err != nil {
-		return
-	}
-
-	offsetForRewards := offset
-
-	defer func() {
-		if withRewards {
-			totals := au.roundTotals[offsetForRewards]
-			proto := au.protos[offsetForRewards]
-			data = data.WithUpdatedRewards(proto, totals.RewardsLevel)
-		}
-	}()
-
-	// Check if this is the most recent round, in which case, we can
-	// use a cache of the most recent account state.
-	if offset == uint64(len(au.deltas)) {
-		macct, ok := au.accounts[addr]
-		if ok {
-			return macct.data, nil
-		}
-	} else {
-		// Check if the account has been updated recently.  Traverse the deltas
-		// backwards to ensure that later updates take priority if present.
-		for offset > 0 {
-			offset--
-			d, ok := au.deltas[offset][addr]
-			if ok {
-				return d.new, nil
-			}
-		}
-	}
-
-	// No updates of this account in the in-memory deltas; use on-disk DB.
-	// The check in roundOffset() made sure the round is exactly the one
-	// present in the on-disk DB.  As an optimization, we avoid creating
-	// a separate transaction here, and directly use a prepared SQL query
-	// against the database.
-	return au.accountsq.lookup(addr)
+	return au.lookup(rnd, addr, withRewards)
```

Comment:
previously we had other callers in the `ledger` package that were calling `lookup` externally. For these refactored, lockless methods, it might be good to change their names like you did with `newBlock` and `newBlockImpl` to make sure we don't accidentally call the wrong version.

Reply:
To add some hopefully helpful commentary, this isn't too different than patterns I've seen in the past like in the ACE C++ library, where xxxx_i methods represented 'internal' methods that assumed a lock was already held by the caller.
This way you knew that a public method, Foo could lock, but (assuming non-reentrant locks) could call no other non _i method.  It could only call _i methods (once obtaining a lock).  _i methods could only call other _i methods.  _i methods could always assume a lock was held.  This could allow more performant non-reentrant locking primitives and avoid deadlocks through a simple but effective stylistic convention.

I think Go's public/private capitalization for this same pattern is a good approach but might break down quickly as it's mixing a valuable, but firm and compiler-enforced visibility pattern with a hint about locking expectations.   Perhaps something similar to the _i)nternal 'hint?'

Reply:
Yes, I agree that we should find a better way to separate external and internal functions. What if `accountUpdates` would have been an interface, and internally, there was a struct which implements it.
Any external references to this interface would be limited ( by definition ) to the interface methods. Internally, we could pick whatever naming convention of our choice, knowing that it won't be exposed outside.
( we've already used this design patten in the agreement code somewhere )